### PR TITLE
[FIRRTL] Use annotations to represent dontTouch

### PR DESCRIPTION
**Pull request by uenoku**
This commit switches back the representation of dontTouch to normal
annotations. We have used symbols to uniformly represent optimization
blocking behavior (#2407), but it has become hard because of the complexity
hierpath updates or field-sensitive symbols. Therefore, this commit
changes the representation of dontTouch to normal annotations.

First, LowerTypes is modified to emit errors for aggregate symbols.
This assumption makes LowerTypes a lot simpler. Actually, symbols are
attached before LowerTypes only by dontTouch so the restriction should
not be a big problem. Second, since we don't have to attach symbols at
LowerTypes, the implementation of `filterAnnotation` is simplified a lot.
Also, the renaming of symbols will not happen anymore so they are removed as well.

### CODE REVIEW

File: test/Dialect/FIRRTL/lower-types.mlir
Comment (uenoku) [inline test/Dialect/FIRRTL/lower-types.mlir:1123]: Can we delete these tests since we have changed the representation of hierpath?

Comment (uenoku) [inline test/Dialect/FIRRTL/lower-types.mlir:1123]: I assumed hierpath with old format will never be created. Is it correct?

File: lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp:261]: This method `applyDontTouch` can be removed and replaced with a complete list of everything that it is allowed to apply to, i.e., `applyWithoutTarget<true, WireOp, NodeOp, MemOp, ...>`.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:552]: Nit: this error message kind of indicates that `LowerTypes` doesn't support lowering symbols when we want to indicate that it is (currently) illegal to pass IR that has symbols into `LowerTypes`. Maybe: `"has a symbol, but no symbols may exist on aggregates passed through LowerTypes"`.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:651]: See better error message suggestion above.

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:1486]: Happy to see this logic gone. 😂

File: test/Dialect/FIRRTL/lower-types.mlir
Comment (seldridge) [reply test/Dialect/FIRRTL/lower-types.mlir:1123]: It could be created by somebody, but not be a "normal" pass pipeline that uses `LowerAnnotations`. It's fine to remove this.

**Review by seldridge — APPROVED** [review]
Nice! I love the dramatic simplification of `LowerTypes`.

This should have an explicit @darthscsi approval before merging. 

Also, can you run this in an internal design and make sure that there are _zero changes_ to anything in the output? (I'll usually check this with `meld` or some graphical diff.)

**Comment by uenoku** [comment]
> Also, can you run this in an internal design and make sure that there are zero changes to anything in the output? (I'll usually check this with meld or some graphical diff.

I'll check, thank you for the suggestion!

**Review by darthscsi — APPROVED** [review]
It's hard to tell, because it's the deleted which shows up, but this does still propagate symbols when they are on fields, right?  (in lowertypes) (symbols being used for stuff besides don't touch)

**Comment by prithayan** [comment]
> It's hard to tell, because it's the deleted which shows up, but this does still propagate symbols when they are on fields, right? (in lowertypes) (symbols being used for stuff besides don't touch)

LowerTypes still does not handle field sensitive fields properly. That should be a different PR for  handling `InnerSymAttr`, after this PR lands.

**Comment by uenoku** [comment]
> LowerTypes still does not handle field sensitive fields properly. That should be a different PR for handling InnerSymAttr, after this PR lands.

Right, LowerTypes is not able to handle field sensitive symbols right now but practically it should be fine for a while because no annotation creates symbols before LowerTypes right now. We can implement it after landing PR.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:171]: Especially since symbols are now no longer "dontTouch", can we separate these?  Are there users expecting the "hasDontTouch" predicate to answer "or has a symbol"?  If so, maybe we can rename this?

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:1486]: Agreed, this is a lot of removed code/complexity!

File: test/Dialect/FIRRTL/annotations.mlir
Comment (dtzSiFive) [inline test/Dialect/FIRRTL/annotations.mlir:742]: Thanks for updating the comment! (and others)

**Review by dtzSiFive — COMMENTED** [review]
Looks good, thanks!

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:171]: I think conventionally we have used a term "dontTouch" to represent an optimization blocking property and inner symbols still imply "dontTouch".

**Comment by uenoku** [comment]
> Also, can you run this in an internal design and make sure that there are zero changes to anything in the output? (I'll usually check this with meld or some graphical diff.

I tested on 4~5 large cores but there is no difference :+1:

**Pull request merged**
